File: solutions/4298-count-smaller-elements-with-opposite-parity/count-smaller-elements-with-opposite-parity.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
    // Standard BIT/Fenwick Tree implementation
    vector<int> bitEven, bitOdd;
    int size;

    void update(vector<int>& bit, int idx, int val) {
        for (; idx <= size; idx += idx & -idx) {
            bit[idx] += val;
        }
    }

    int query(vector<int>& bit, int idx) {
        int sum = 0;
        for (; idx > 0; idx -= idx & -idx) {
            sum += bit[idx];
        }
        return sum;
    }

public:
    vector<int> countSmallerOppositeParity(vector<int>& nums) {
        int n = nums.size();
        if (n == 0) return {};

        // 1. Coordinate Compression
        // We need to map large values (up to 10^9) to a small range (1 to n)
        vector<int> sorted_nums = nums;
        sort(sorted_nums.begin(), sorted_nums.end());
        sorted_nums.erase(unique(sorted_nums.begin(), sorted_nums.end()), sorted_nums.end());
        
        size = sorted_nums.size();
        bitEven.assign(size + 1, 0);
        bitOdd.assign(size + 1, 0);

        vector<int> answer(n);

        // 2. Process from right to left
        for (int i = n - 1; i >= 0; --i) {
            // Find the rank (1-indexed position) of nums[i]
            int rank = lower_bound(sorted_nums.begin(), sorted_nums.end(), nums[i]) - sorted_nums.begin() + 1;
            
            if (nums[i] % 2 == 0) {
                // If current is even, count smaller odds seen so far
                answer[i] = query(bitOdd, rank - 1);
                update(bitEven, rank, 1);
            } else {
                // If current is odd, count smaller evens seen so far
                answer[i] = query(bitEven, rank - 1);
                update(bitOdd, rank, 1);
            }
        }

        return answer;
    }
};

File: solutions/4298-count-smaller-elements-with-opposite-parity/count-smaller-elements-with-opposite-parity.java
class Solution {
    // time = O(nlogn), space = O(n)
    public int[] countSmallerOppositeParity(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        int[] b = nums.clone();
        Arrays.sort(b);
        HashMap<Integer, Integer> map = new HashMap<>();
        int m = 0;
        for (int x : b) {
            if (!map.containsKey(x)) map.put(x, m++);
        }
        
        Fenwick fen0 = new Fenwick(m);
        Fenwick fen1 = new Fenwick(m);
        for (int i = n - 1; i >= 0; i--) {
            int v = nums[i], idx = map.get(v), p = v % 2;
            if (p == 0) {
                res[i] = fen1.sum(idx);
                fen0.add(idx, 1);
            } else {
                res[i] = fen0.sum(idx);
                fen1.add(idx, 1);
            }
        }
        return res;
    }

    class Fenwick {
        private int n;
        private int[] a;
        public Fenwick(int n) {
            this.n = n;
            this.a = new int[n + 1];
        }

        private void add(int x, int v) {
            for (int i = x + 1; i <= n; i += i & -i) {
                a[i - 1] = a[i - 1] + v;
            }
        }

        private int sum(int x) {
            int ans = 0;
            for (int i = x; i > 0; i -= i & -i) {
                ans = ans + a[i - 1];
            }
            return ans;
        }

        private int rangeSum(int l, int r) { // 左开右闭
            return sum(r) - sum(l);
        }
    }
}